Spigot-Server-Patches/0236-Implemented-the-cancellable-ban-events.patch

```diff
@@ -0,0 +1,131 @@
+ 
+     @Override
+     public org.bukkit.BanEntry addBan(String target, String reason, Date expires, String source) {
++
+         Validate.notNull(target, "Ban target cannot be null");
+ 
+         IpBanEntry entry = new IpBanEntry(target, new Date(),
+                 StringUtils.isBlank(source) ? null : source, expires,
+                 StringUtils.isBlank(reason) ? null : reason);
+ 
+-        list.add(entry);
++        BanIpEvent event = new BanIpEvent(new CraftIpBanEntry(target, entry, list));
++        event = CraftEventFactory.callEvent(event);
++
++        if(event.isCancelled()) return null;
++        IpBanEntry newEntry =  new IpBanEntry(event.getTarget(), event.getCreated(), event.getSource(), event.getExpiration(), event.getReason());
++
++        list.add(newEntry);
+ 
+         try {
+             list.save();
+@@ -50,7 +57,7 @@ public class CraftIpBanList implements org.bukkit.BanList {
+             Bukkit.getLogger().log(Level.SEVERE, "Failed to save banned-ips.json, {0}", ex.getMessage());
+         }
+ 
+-        return new CraftIpBanEntry(target, entry, list);
++        return new CraftIpBanEntry(event.getTarget(), newEntry, list);
+     }
+ 
+     @Override
+diff --git a/src/main/java/org/bukkit/craftbukkit/CraftProfileBanList.java b/src/main/java/org/bukkit/craftbukkit/CraftProfileBanList.java
+index 84a1f9c8..2ac9db43 100644
+--- a/src/main/java/org/bukkit/craftbukkit/CraftProfileBanList.java
++++ b/src/main/java/org/bukkit/craftbukkit/CraftProfileBanList.java
+@@ -1,21 +1,21 @@
+ package org.bukkit.craftbukkit;
+ 
+-import java.io.IOException;
+-import java.util.Date;
+-import java.util.Set;
+-
++import com.google.common.collect.ImmutableSet;
++import com.mojang.authlib.GameProfile;
+ import net.minecraft.server.GameProfileBanEntry;
+ import net.minecraft.server.GameProfileBanList;
+ import net.minecraft.server.JsonListEntry;
+ import net.minecraft.server.MinecraftServer;
+-
+ import org.apache.commons.lang.StringUtils;
+ import org.apache.commons.lang.Validate;
++import org.bukkit.Bukkit;
++import org.bukkit.craftbukkit.event.CraftEventFactory;
++import org.bukkit.event.server.BanProfileEvent;
+ 
+-import com.google.common.collect.ImmutableSet;
+-import com.mojang.authlib.GameProfile;
++import java.io.IOException;
++import java.util.Date;
++import java.util.Set;
+ import java.util.logging.Level;
+-import org.bukkit.Bukkit;
+ 
+ public class CraftProfileBanList implements org.bukkit.BanList {
+     private final GameProfileBanList list;
+@@ -54,7 +54,14 @@ public class CraftProfileBanList implements org.bukkit.BanList {
+                 StringUtils.isBlank(source) ? null : source, expires,
+                 StringUtils.isBlank(reason) ? null : reason);
+ 
+-        list.add(entry);
++        BanProfileEvent event = new BanProfileEvent(new CraftProfileBanEntry(profile, entry, list));
++        event = CraftEventFactory.callEvent(event);
++        if(event.isCancelled()) return null;
```

Comment:
I am partially concerned with the practices in this PR, this really isn't all that inclusive either due to how CB has wrapped this, in the event that a ban is done with the vanilla command, this won't actually be caught by this PR.

I would say that these changes are actually better being made in the NMS BanList's, that way it's more inclusive